// @ant-design/flowchart/lib/components/editor-panel/form-panel.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FormPanel = exports.CONTAINER_CLASS = void 0;
var react_1 = __importStar(require("react"));
// import { JsonSchemaForm } from '@antv/xflow';
var xflow_1 = require("@antv/xflow");
var icons_1 = require("@ant-design/icons");
var canvas_json_schema_form_1 = require("../canvas-json-schema-form");
var form_schema_service_1 = require("./form-schema-service");
var control_map_service_1 = require("./control-map-service");
exports.CONTAINER_CLASS = 'flowchart-editor-panel-collpase';
var FormPanel = function (props) {
    var _a = props.controlMapService, controlMapService = _a === void 0 ? control_map_service_1.defaultControlMapService : _a, _b = props.formSchemaService, formSchemaService = _b === void 0 ? form_schema_service_1.defaultFormSchemaService : _b, _c = props.position, position = _c === void 0 ? { width: 240, top: 0, bottom: 0, right: 0 } : _c, _d = props.show, show = _d === void 0 ? true : _d, rest = __rest(props, ["controlMapService", "formSchemaService", "position", "show"]);
    if (!show) {
        return null;
    }
    var _e = position.width, width = _e === void 0 ? 200 : _e, right = position.right;
    var _f = (0, react_1.useState)(false), collpased = _f[0], setCollpased = _f[1];
    return (react_1.default.createElement(xflow_1.WorkspacePanel, { className: exports.CONTAINER_CLASS, position: __assign(__assign({}, position), { right: !collpased ? right : -width }) },
        react_1.default.createElement("div", { className: "".concat(exports.CONTAINER_CLASS, "-wrapper") },
            react_1.default.createElement(canvas_json_schema_form_1.JsonSchemaForm, __assign({ targetType: ['node', 'edge', 'canvas', 'group'], controlMapService: controlMapService, formSchemaService: formSchemaService, header: null, position: __assign(__assign({}, position), { top: 0 }), prefixClz: "charts-form-editor" }, rest)),
            react_1.default.createElement("div", { className: "".concat(exports.CONTAINER_CLASS, "-icon"), style: {
                    top: 21,
                    left: !collpased ? -10 : -20,
                    borderRadius: !collpased ? '50%' : '50% 0 0  50%',
                    borderRight: !collpased ? '' : 'none',
                }, onClick: function () {
                    setCollpased(!collpased);
                } }, collpased ? react_1.default.createElement(icons_1.DoubleLeftOutlined, null) : react_1.default.createElement(icons_1.DoubleRightOutlined, null)))));
};
exports.FormPanel = FormPanel;
