// @ant-design/flowchart/lib/components/canvas-json-schema-form/schema-form/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SchemaForm = exports.FormItemWrapper = void 0;
var react_1 = __importStar(require("react"));
var antd_1 = require("antd");
var classnames_1 = __importDefault(require("classnames"));
var util_1 = require("../../../util");
var index_1 = require("./control-map/index");
var form_item_wrapper_1 = require("./common/form-item-wrapper");
Object.defineProperty(exports, "FormItemWrapper", { enumerable: true, get: function () { return form_item_wrapper_1.FormItemWrapper; } });
var TabPane = antd_1.Tabs.TabPane;
var SchemaForm = function (props) {
    var _a;
    var _b;
    var schema = props.schema, formInstance = props.form, className = props.className, triggerUpdate = props.triggerUpdate, defaultControlRender = props.defaultControlRender, afterUpdatingCb = props.afterUpdatingCb, controlMapfromParent = props.controlMap, otherProps = __rest(props, ["schema", "form", "className", "triggerUpdate", "defaultControlRender", "afterUpdatingCb", "controlMap"]);
    var tabs = schema.tabs;
    var form = antd_1.Form.useForm(formInstance)[0];
    var controlMap = react_1.default.useMemo(function () { return (props.controlMap ? props.controlMap : (0, index_1.makeControlMap)(index_1.xflowDefaultControls)); }, [props.controlMap]);
    var innerTriggerUpdate = (0, react_1.useCallback)(function (values) {
        if (typeof triggerUpdate === 'function') {
            triggerUpdate(form, values);
        }
    }, [form, triggerUpdate]);
    /** 初次挂载时设置一次值以激活首次联动 */
    (0, react_1.useEffect)(function () {
        form.setFieldsValue({});
    }, [form]);
    /** 切换标签时激活联动 */
    var onClickTab = (0, react_1.useCallback)(function () {
        requestAnimationFrame(function () {
            form.setFieldsValue({});
        });
    }, [form]);
    return (react_1.default.createElement(antd_1.Form, __assign({ size: "small", form: form, layout: "vertical", className: (0, classnames_1.default)('xflow-json-form', className) }, otherProps),
        react_1.default.createElement(antd_1.Tabs, { type: "card", defaultActiveKey: (_b = tabs[0]) === null || _b === void 0 ? void 0 : _b.name, className: (0, classnames_1.default)((_a = {},
                _a['tabs'] = true,
                _a['xTab'] = true,
                _a['singleTab'] = (tabs === null || tabs === void 0 ? void 0 : tabs.length) === 1,
                _a['coupleTab'] = (tabs === null || tabs === void 0 ? void 0 : tabs.length) === 2,
                _a['ternateTab'] = (tabs === null || tabs === void 0 ? void 0 : tabs.length) === 3,
                _a)), onTabClick: onClickTab }, tabs.map(function (tab) {
            var tabName = tab.name, groups = tab.groups;
            return (react_1.default.createElement(TabPane, { key: tabName, tab: tabName },
                groups.length === 0 && react_1.default.createElement(antd_1.Empty, null),
                groups.length > 0 &&
                    groups.map(function (group) {
                        var _a = group.controls, groupControls = _a === void 0 ? [] : _a;
                        if (groupControls.length === 0) {
                            return react_1.default.createElement(antd_1.Empty, { style: { padding: '12px 0' } });
                        }
                        return groupControls.map(function (control) {
                            var shape = control.shape, controlName = control.name;
                            var ControlComponent = (0, index_1.getControlFromMap)(shape, controlMap, defaultControlRender);
                            if (!ControlComponent) {
                                util_1.Log.error('未找到对应的控件:', shape);
                                return null;
                            }
                            return (react_1.default.createElement(ControlComponent, { key: controlName, form: form, controlSchema: control, triggerUpdate: innerTriggerUpdate, afterUpdatingCb: afterUpdatingCb }));
                        });
                    })));
        }))));
};
exports.SchemaForm = SchemaForm;
