// @ant-design/flowchart/lib/components/editor-panel/control-map-service/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CanvasService = exports.GroupService = exports.EdgeService = exports.NodeService = exports.EditorPanels = exports.defaultControlMapService = void 0;
var node_1 = require("./components/node");
Object.defineProperty(exports, "NodeService", { enumerable: true, get: function () { return node_1.NodeService; } });
var edge_1 = require("./components/edge");
Object.defineProperty(exports, "EdgeService", { enumerable: true, get: function () { return edge_1.EdgeService; } });
var group_1 = require("./components/group");
Object.defineProperty(exports, "GroupService", { enumerable: true, get: function () { return group_1.GroupService; } });
var canvas_1 = require("./components/canvas");
Object.defineProperty(exports, "CanvasService", { enumerable: true, get: function () { return canvas_1.CanvasService; } });
var fields_1 = require("./components/fields");
Object.defineProperty(exports, "EditorPanels", { enumerable: true, get: function () { return fields_1.EditorPanels; } });
/** 默认支持修改标签和重命名功能 */
var defaultControlMapService = function (controlMap) {
    controlMap.set('node-service', node_1.NodeService);
    controlMap.set('edge-service', edge_1.EdgeService);
    controlMap.set('group-service', group_1.GroupService);
    controlMap.set('canvas-service', canvas_1.CanvasService);
    return controlMap;
};
exports.defaultControlMapService = defaultControlMapService;
