// @ant-design/graphs/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.G6 = exports.RadialGraph = exports.OrganizationalGraph = exports.IndentedTreeGraph = exports.DagreFundFlowGraph = exports.IndentedTree = exports.DagreGraph = exports.OrganizationTreeGraph = exports.FundFlowGraph = exports.OrganizationGraph = exports.DecompositionTreeGraph = exports.RadialTreeGraph = exports.FlowAnalysisGraph = void 0;
var g6_1 = __importDefault(require("@antv/g6"));
exports.G6 = g6_1.default;
var organizationGraph_1 = __importDefault(require("./components/organizationGraph"));
exports.OrganizationGraph = organizationGraph_1.default;
var radialTreeGraph_1 = __importDefault(require("./components/radialTreeGraph"));
exports.RadialTreeGraph = radialTreeGraph_1.default;
var flowAnalysisGraph_1 = __importDefault(require("./components/flowAnalysisGraph"));
exports.FlowAnalysisGraph = flowAnalysisGraph_1.default;
var decompositionTreeGraph_1 = __importDefault(require("./components/decompositionTreeGraph"));
exports.DecompositionTreeGraph = decompositionTreeGraph_1.default;
var fundFlowGraph_1 = __importDefault(require("./components/fundFlowGraph"));
exports.FundFlowGraph = fundFlowGraph_1.default;
/** @deprecated start */
var graph_1 = require("./obsolescent/graph");
Object.defineProperty(exports, "OrganizationTreeGraph", { enumerable: true, get: function () { return graph_1.OrganizationTreeGraph; } });
Object.defineProperty(exports, "DagreGraph", { enumerable: true, get: function () { return graph_1.DagreGraph; } });
Object.defineProperty(exports, "IndentedTree", { enumerable: true, get: function () { return graph_1.IndentedTree; } });
Object.defineProperty(exports, "DagreFundFlowGraph", { enumerable: true, get: function () { return graph_1.DagreFundFlowGraph; } });
var graphs_1 = require("./obsolescent/graphs");
Object.defineProperty(exports, "IndentedTreeGraph", { enumerable: true, get: function () { return graphs_1.IndentedTreeGraph; } });
Object.defineProperty(exports, "OrganizationalGraph", { enumerable: true, get: function () { return graphs_1.OrganizationalGraph; } });
Object.defineProperty(exports, "RadialGraph", { enumerable: true, get: function () { return graphs_1.RadialGraph; } });
__exportStar(require("./interface"), exports);
