// @antv/l7-map/lib/handler/click_zoom.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var ClickZoomHandler = function () {
  function ClickZoomHandler() {
    (0, _classCallCheck2.default)(this, ClickZoomHandler);
    (0, _defineProperty2.default)(this, "enabled", void 0);
    (0, _defineProperty2.default)(this, "active", void 0);
    this.reset();
  }

  (0, _createClass2.default)(ClickZoomHandler, [{
    key: "reset",
    value: function reset() {
      this.active = false;
    }
  }, {
    key: "dblclick",
    value: function dblclick(e, point) {
      e.preventDefault();
      return {
        cameraAnimation: function cameraAnimation(map) {
          map.easeTo({
            duration: 300,
            zoom: map.getZoom() + (e.shiftKey ? -1 : 1),
            around: map.unproject(point)
          }, {
            originalEvent: e
          });
        }
      };
    }
  }, {
    key: "enable",
    value: function enable() {
      this.enabled = true;
    }
  }, {
    key: "disable",
    value: function disable() {
      this.enabled = false;
      this.reset();
    }
  }, {
    key: "isEnabled",
    value: function isEnabled() {
      return this.enabled;
    }
  }, {
    key: "isActive",
    value: function isActive() {
      return this.active;
    }
  }]);
  return ClickZoomHandler;
}();

exports.default = ClickZoomHandler;
