// @ant-design/flowchart/es/components/group-panel/group.js
import React from 'react';
import { PlusSquareOutlined, MinusSquareOutlined } from '@ant-design/icons';
import { useXFlowApp, XFlowGroupCommands } from '@antv/xflow';
export var GroupNode = function (props) {
    var cell = props.cell, _a = props.data, label = _a.label, stroke = _a.stroke, fill = _a.fill, fontSize = _a.fontSize, fontFill = _a.fontFill, _b = _a.width, width = _b === void 0 ? 200 : _b, _c = _a.isCollapsed, isCollapsed = _c === void 0 ? false : _c;
    var app = useXFlowApp();
    // const isCollapse = cell.getProp('isCollapsed') || false;
    var onExpand = function (e) {
        app.executeCommand(XFlowGroupCommands.COLLAPSE_GROUP.id, {
            nodeId: cell.id,
            isCollapsed: false,
            collapsedSize: { width: width, height: 40 },
        });
    };
    var onCollapse = function (e) {
        app.executeCommand(XFlowGroupCommands.COLLAPSE_GROUP.id, {
            nodeId: cell.id,
            isCollapsed: true,
            collapsedSize: { width: width, height: 40 },
            gap: 3,
        });
    };
    return (React.createElement("div", { className: "xflow-group-node", style: {
            borderColor: stroke,
            backgroundColor: fill,
            fontSize: fontSize,
            color: fontFill,
        } },
        React.createElement("div", { className: "xflow-group-header" },
            React.createElement("div", { className: "header-left" }, label),
            React.createElement("div", { className: "header-right" },
                isCollapsed && React.createElement(PlusSquareOutlined, { onClick: onExpand }),
                !isCollapsed && React.createElement(MinusSquareOutlined, { onClick: onCollapse })))));
};
