// @ant-design/flowchart/es/components/canvas-toolbar/config.js
import { uuidv4 } from '@antv/xflow';
export var CONFIG_TYPE = 'CanvasToolbar';
var ToolbarConfig = /** @class */ (function () {
    /** uuid */
    function ToolbarConfig() {
        var _this = this;
        /** config type */
        this.CONFIG_TYPE = CONFIG_TYPE;
        /** 设置toolbar id */
        this.setToolbarName = function (moduleId) {
            _this.moduleId = CONFIG_TYPE + '__' + moduleId;
        };
        /** 通过订阅 modelService 设置toolbar model */
        this.setToolbarModelService = function (service) {
            _this.toolbarModelService = service;
        };
        /** 设置组件的props */
        this.setCustomToolbarRender = function (service) {
            _this.customRenderService = service;
        };
        /** 获取Props */
        this.getConfig = function () {
            return {
                CONFIG_TYPE: _this.CONFIG_TYPE,
                toolbarCustomRender: _this.customRenderService,
                toolbarModelService: _this.toolbarModelService,
            };
        };
        this.moduleId = CONFIG_TYPE + '__' + uuidv4();
    }
    /** dispose */
    ToolbarConfig.prototype.dispose = function () { };
    return ToolbarConfig;
}());
export { ToolbarConfig };
