// @antv/l7-layers/lib/imagetile/utils/tileCache.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _lruCache = _interopRequireDefault(require("./lruCache"));

var TileCache = function () {
  function TileCache() {
    var limit = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 50;
    var tileDestroy = arguments.length > 1 ? arguments[1] : undefined;
    (0, _classCallCheck2.default)(this, TileCache);
    (0, _defineProperty2.default)(this, "cache", void 0);
    this.cache = new _lruCache.default(limit, tileDestroy);
  }

  (0, _createClass2.default)(TileCache, [{
    key: "getTile",
    value: function getTile(key) {
      return this.cache.get(key);
    }
  }, {
    key: "setTile",
    value: function setTile(tile, key) {
      this.cache.set(key, tile);
    }
  }, {
    key: "destory",
    value: function destory() {
      this.cache.clear();
    }
  }, {
    key: "setNeedUpdate",
    value: function setNeedUpdate() {
      var _this = this;

      this.cache.order.forEach(function (key) {
        var tile = _this.cache.get(key);

        tile.needUpdate = true;
      });
    }
  }]);
  return TileCache;
}();

exports.default = TileCache;
