// @antv/xflow-extension/lib/canvas-dag-extension/contributions/dag.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DagHooksContribution = exports.getDagOptions = exports.ANT_PREFIX = exports.DAG_DEFAULT_CONIFG = exports.NsAddEdgeEvent = void 0;
var tslib_1 = require("tslib");
var xflow_core_1 = require("@antv/xflow-core");
var xflow_core_2 = require("@antv/xflow-core");
var xflow_hook_1 = require("@antv/xflow-hook");
var node_1 = require("../x6-extension/node");
var edge_1 = require("../x6-extension/edge");
var constants_1 = require("../constants");
var command_1 = require("./command");
var interface_1 = require("../interface");
var interface_2 = require("../interface");
var NsAddEdgeEvent;
(function (NsAddEdgeEvent) {
    NsAddEdgeEvent.EVENT_NAME = 'ADD_EDGE_CMD_EVENT';
})(NsAddEdgeEvent = exports.NsAddEdgeEvent || (exports.NsAddEdgeEvent = {}));
var DAG_DEFAULT_CONIFG;
(function (DAG_DEFAULT_CONIFG) {
    DAG_DEFAULT_CONIFG.router = { name: 'manhattan' };
    DAG_DEFAULT_CONIFG.connector = {
        name: 'rounded',
        args: { radius: 15 },
    };
})(DAG_DEFAULT_CONIFG = exports.DAG_DEFAULT_CONIFG || (exports.DAG_DEFAULT_CONIFG = {}));
exports.ANT_PREFIX = 'ant';
var getDagOptions = function (props) {
    var _a = props.layout, layout = _a === void 0 ? interface_1.LayoutEnum.TOP_BOTTOM : _a, _b = props.router, router = _b === void 0 ? DAG_DEFAULT_CONIFG.router : _b, _c = props.connector, connector = _c === void 0 ? DAG_DEFAULT_CONIFG.connector : _c;
    var targetPortType = layout === interface_1.LayoutEnum.TOP_BOTTOM ? xflow_core_1.NsGraph.AnchorGroup.TOP : xflow_core_1.NsGraph.AnchorGroup.LEFT;
    var dagOptions = {
        grid: false,
        keyboard: {
            enabled: true,
        },
        // 点选/框选配置（详细文档：https://X6.antv.vision/zh/docs/tutorial/basic/selection）
        selecting: {
            enabled: true,
            multiple: true,
            selectCellOnMoved: true,
            showNodeSelectionBox: false,
            // 框选可以选中edge
            rubberEdge: true,
            // 框选可以选中node
            rubberNode: true,
            movable: true,
        },
        connecting: {
            //链接桩的位置 https://x6.antv.vision/zh/docs/api/registry/node-anchor
            sourceAnchor: layout === interface_1.LayoutEnum.TOP_BOTTOM ? 'bottom' : 'right',
            //链接桩的位置 https://x6.antv.vision/zh/docs/api/registry/node-anchor
            targetAnchor: layout === interface_1.LayoutEnum.TOP_BOTTOM ? 'center' : 'left',
            connectionPoint: 'anchor',
            snap: { radius: 20 },
            router: router,
            connector: connector,
            highlight: true,
            dangling: false,
            createEdge: function () {
                /* eslint-disable-next-line @typescript-eslint/no-this-alias */
                var graph = this;
                var edge = new edge_1.XFlowEdge({
                    attrs: {
                        line: {
                            strokeDasharray: '5 5',
                            stroke: '#888',
                            strokeWidth: 1,
                            targetMarker: {
                                name: 'block',
                                args: {
                                    size: '6',
                                },
                            },
                        },
                    },
                });
                var addEdge = function (args) {
                    var isNew = args.isNew;
                    var edgeCell = args.edge;
                    /** 没有edge:connected时，会导致graph.once的事件没有执行 */
                    if (isNew && edgeCell.isEdge() && edgeCell === edge) {
                        var portId = edgeCell.getTargetPortId();
                        var targetNode = edgeCell.getTargetCell();
                        if (targetNode && targetNode.isNode()) {
                            targetNode.setPortProp(portId, 'connected', false);
                            edgeCell.attr({
                                line: {
                                    strokeDasharray: '',
                                    targetMarker: '',
                                    stroke: '#d5d5d5',
                                },
                            });
                            var targetPortId = edgeCell.getTargetPortId();
                            var sourcePortId = edgeCell.getSourcePortId();
                            var sourceCellId = edgeCell.getSourceCellId();
                            var targetCellId = edgeCell.getTargetCellId();
                            graph.trigger(NsAddEdgeEvent.EVENT_NAME, {
                                targetPortId: targetPortId,
                                sourcePortId: sourcePortId,
                                source: sourceCellId,
                                target: targetCellId,
                                edge: edge,
                            });
                        }
                    }
                };
                graph.once('edge:connected', addEdge);
                return edge;
            },
            validateEdge: function (args) {
                var _a;
                var edge = args.edge;
                return !!((_a = edge === null || edge === void 0 ? void 0 : edge.target) === null || _a === void 0 ? void 0 : _a.port);
            },
            // 是否触发交互事件
            validateMagnet: function (_a) {
                var magnet = _a.magnet;
                return magnet.getAttribute('port-group') !== targetPortType;
            },
            // 显示可用的链接桩
            validateConnection: function (_a) {
                var sourceView = _a.sourceView, targetView = _a.targetView, sourceMagnet = _a.sourceMagnet, targetMagnet = _a.targetMagnet;
                // 不允许连接到自己
                if (sourceView === targetView) {
                    return false;
                }
                // 只能从上游节点的输出链接桩创建连接
                if (!sourceMagnet || sourceMagnet.getAttribute('port-group') === targetPortType) {
                    return false;
                }
                // 只能连接到下游节点的输入桩
                if (!targetMagnet || targetMagnet.getAttribute('port-group') !== targetPortType) {
                    return false;
                }
                var node = targetView.cell;
                // 判断目标链接桩是否可连接
                var portId = targetMagnet.getAttribute('port');
                var port = node.getPort(portId);
                return !(port && port.connected);
            },
        },
        highlighting: {
            nodeAvailable: {
                name: 'className',
                args: {
                    className: 'available',
                },
            },
            magnetAvailable: {
                name: 'className',
                args: {
                    className: 'available',
                },
            },
            magnetAdsorbed: {
                name: 'className',
                args: {
                    className: 'adsorbed',
                },
            },
        },
        scaling: {
            max: 1.05,
            min: 0.01,
        },
    };
    return dagOptions;
};
exports.getDagOptions = getDagOptions;
/**
 * 内置的hook contribution
 * 处理 config上的runtime的注册项
 */
var DagHooksContribution = /** @class */ (function () {
    function DagHooksContribution() {
        var _this = this;
        /** 注册Command */
        this.registerGraphCommands = function (commands) {
            commands.registerCommand(command_1.NsGraphStatusCommand.command, {
                createCommand: _this.commandFactory,
            });
        };
        /** 注册Hub */
        this.toDispose = new xflow_core_1.DisposableCollection();
        /** 注册Hook */
        this.registerHook = function (hooks) { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
            var toDispose, disposables;
            var _this = this;
            return (0, tslib_1.__generator)(this, function (_a) {
                toDispose = new xflow_core_1.DisposableCollection();
                disposables = [
                    hooks.addNode.registerHook({
                        name: 'dag-add-node',
                        handler: function (args) { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
                            var cellFactory;
                            var _this = this;
                            return (0, tslib_1.__generator)(this, function (_a) {
                                cellFactory = function (nodeConfig) { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
                                    var node;
                                    return (0, tslib_1.__generator)(this, function (_a) {
                                        node = new node_1.XFlowNode((0, tslib_1.__assign)({}, nodeConfig));
                                        return [2 /*return*/, node];
                                    });
                                }); };
                                args.cellFactory = cellFactory;
                                return [2 /*return*/];
                            });
                        }); },
                    }),
                    hooks.addEdge.registerHook({
                        name: 'dag-add-edge',
                        handler: function (args) { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
                            var _a, layout, cellFactory;
                            var _this = this;
                            return (0, tslib_1.__generator)(this, function (_b) {
                                switch (_b.label) {
                                    case 0: return [4 /*yield*/, this.propConfig.getConfig()];
                                    case 1:
                                        _a = (_b.sent()).layout, layout = _a === void 0 ? interface_1.LayoutEnum.TOP_BOTTOM : _a;
                                        cellFactory = function (edgeConfig) { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
                                            var cell;
                                            return (0, tslib_1.__generator)(this, function (_a) {
                                                cell = new edge_1.XFlowEdge((0, tslib_1.__assign)((0, tslib_1.__assign)({}, edgeConfig), { id: edgeConfig.id, source: {
                                                        cell: edgeConfig.source,
                                                        port: edgeConfig.sourcePortId,
                                                        anchor: {
                                                            name: layout === interface_1.LayoutEnum.TOP_BOTTOM ? 'bottom' : 'right',
                                                        },
                                                    }, target: {
                                                        cell: edgeConfig.target,
                                                        port: edgeConfig.targetPortId,
                                                        anchor: {
                                                            name: layout === interface_1.LayoutEnum.TOP_BOTTOM ? 'center' : 'right',
                                                        },
                                                    }, attrs: {
                                                        line: {
                                                            strokeDasharray: '',
                                                            targetMarker: '',
                                                            stroke: '#d5d5d5',
                                                            strokeWidth: 1,
                                                        },
                                                    }, data: (0, tslib_1.__assign)({}, edgeConfig) }));
                                                return [2 /*return*/, cell];
                                            });
                                        }); };
                                        args.cellFactory = cellFactory;
                                        return [2 /*return*/];
                                }
                            });
                        }); },
                    }),
                    hooks.addEdge.registerHook({
                        name: 'after add edge, set target port props',
                        handler: function (handlerArgs, handler) { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
                            var main;
                            var _this = this;
                            return (0, tslib_1.__generator)(this, function (_a) {
                                main = function (args) { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
                                    var res, targetNode, portId;
                                    return (0, tslib_1.__generator)(this, function (_a) {
                                        switch (_a.label) {
                                            case 0: return [4 /*yield*/, handler(args)];
                                            case 1:
                                                res = _a.sent();
                                                if (res && res.edgeCell) {
                                                    targetNode = res.edgeCell.getTargetCell();
                                                    portId = res.edgeCell.getTargetPortId();
                                                    targetNode.setPortProp(portId, 'connected', true);
                                                }
                                                return [2 /*return*/, res];
                                        }
                                    });
                                }); };
                                return [2 /*return*/, main];
                            });
                        }); },
                    }),
                    hooks.delEdge.registerHook({
                        name: 'afetr del edge, reset target node port props',
                        handler: function (args, handler) { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
                            var newHandler;
                            var _this = this;
                            return (0, tslib_1.__generator)(this, function (_a) {
                                newHandler = function (handlerArgs) { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
                                    var result, targetCell, targetPortId;
                                    return (0, tslib_1.__generator)(this, function (_a) {
                                        switch (_a.label) {
                                            case 0: return [4 /*yield*/, handler(handlerArgs)];
                                            case 1:
                                                result = _a.sent();
                                                targetCell = result.targetCell, targetPortId = result.targetPortId;
                                                if (targetCell && targetCell.isNode() && targetPortId) {
                                                    targetCell.setPortProp(targetPortId, 'connected', false);
                                                }
                                                return [2 /*return*/, result];
                                        }
                                    });
                                }); };
                                return [2 /*return*/, newHandler];
                            });
                        }); },
                    }),
                    hooks.afterGraphInit.registerHook({
                        name: 'call add edge to replace temp edge',
                        handler: function (handlerArgs) { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
                            var commandService, graph;
                            return (0, tslib_1.__generator)(this, function (_a) {
                                commandService = handlerArgs.commandService, graph = handlerArgs.graph;
                                graph.on(NsAddEdgeEvent.EVENT_NAME, function (args) {
                                    commandService.executeCommand(xflow_core_1.XFlowEdgeCommands.ADD_EDGE.id, { edgeConfig: args });
                                    args.edge.remove();
                                });
                                return [2 /*return*/];
                            });
                        }); },
                    }),
                    hooks.graphOptions.registerHook({
                        name: 'dag-extension-x6-options',
                        handler: function (args) { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
                            var props, dagOptions;
                            return (0, tslib_1.__generator)(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, this.propConfig.getConfig()];
                                    case 1:
                                        props = _a.sent();
                                        dagOptions = (0, exports.getDagOptions)(props);
                                        Object.assign(args, dagOptions);
                                        return [2 /*return*/];
                                }
                            });
                        }); },
                    }),
                ];
                toDispose.pushAll(disposables);
                return [2 /*return*/, xflow_core_1.Disposable.create(function () { })];
            });
        }); };
        /** 注册Hub */
        this.registerHookHub = function (registry) { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
            return (0, tslib_1.__generator)(this, function (_a) {
                return [2 /*return*/, registry.registerHookHub(command_1.NsGraphStatusCommand.hookKey, new xflow_hook_1.HookHub())];
            });
        }); };
    }
    /** 扩展Model */
    DagHooksContribution.prototype.registerModel = function (registry) {
        var _this = this;
        /** node status map */
        registry.registerModel({
            id: constants_1.GRAPH_STATUS_INFO.id,
            getInitialValue: function () { return ({
                statusMap: new Map(),
                subscription: new xflow_core_1.DisposableCollection(),
                graphStatus: constants_1.GRAPH_STATUS_INFO.StatusEnum.DEFAULT,
            }); },
            watchChange: function (self) { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
                var subscription;
                return (0, tslib_1.__generator)(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, self.getValidValue()];
                        case 1:
                            subscription = (_a.sent()).subscription;
                            return [2 /*return*/, subscription];
                    }
                });
            }); },
        });
    };
    (0, tslib_1.__decorate)([
        xflow_core_1.ManaSyringe.inject(xflow_core_2.IGraphCommandFactory),
        (0, tslib_1.__metadata)("design:type", Function)
    ], DagHooksContribution.prototype, "commandFactory", void 0);
    (0, tslib_1.__decorate)([
        xflow_core_1.ManaSyringe.inject(interface_2.IComponentConfig),
        (0, tslib_1.__metadata)("design:type", Object)
    ], DagHooksContribution.prototype, "propConfig", void 0);
    DagHooksContribution = (0, tslib_1.__decorate)([
        xflow_core_1.ManaSyringe.singleton({
            contrib: [xflow_core_2.IHookContribution, xflow_core_2.IModelContribution, xflow_core_2.IGraphCommandContribution],
        })
    ], DagHooksContribution);
    return DagHooksContribution;
}());
exports.DagHooksContribution = DagHooksContribution;
