// @antv/xflow-extension/lib/flowchart-node-panel/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createPath = exports.setNodeRender = exports.nodeService = exports.getRegisterNode = void 0;
var tslib_1 = require("tslib");
var xflow_core_1 = require("@antv/xflow-core");
var lodash_1 = require("lodash");
var NodesComponent = (0, tslib_1.__importStar)(require("./nodes"));
var constants_1 = require("./constants");
/** 和 graph config 注册的节点保持一致 */
var getAnchorStyle = function (position) {
    return {
        position: { name: position },
        attrs: {
            circle: {
                r: 4,
                magnet: true,
                stroke: '#31d0c6',
                strokeWidth: 2,
                fill: '#fff',
                style: {
                    visibility: 'hidden',
                },
            },
        },
        zIndex: 10,
    };
};
var getPorts = function (position) {
    if (position === void 0) { position = ['top', 'right', 'bottom', 'left']; }
    return {
        items: position.map(function (name) {
            return { group: name, id: (0, xflow_core_1.uuidv4)() };
        }),
        groups: {
            top: getAnchorStyle('top'),
            right: getAnchorStyle('right'),
            bottom: getAnchorStyle('bottom'),
            left: getAnchorStyle('left'),
        },
    };
};
var getRegisterNode = function (nodes) {
    return (nodes || []).map(function (item) {
        var name = item.name, popover = item.popover, _a = item.label, label = _a === void 0 ? '' : _a, _b = item.width, width = _b === void 0 ? constants_1.NODE_HEIGHT : _b, _c = item.height, height = _c === void 0 ? constants_1.NODE_HEIGHT : _c, ports = item.ports;
        var id = (0, xflow_core_1.uuidv4)(); // 暂不使用上层数据
        return {
            id: id,
            renderKey: name,
            name: name,
            label: label,
            popoverContent: popover,
            width: width,
            height: height,
            ports: ports || getPorts(),
            originData: (0, tslib_1.__assign)({}, item),
            isCustom: true,
        };
    });
};
exports.getRegisterNode = getRegisterNode;
var nodeService = function (nodes) { return (0, tslib_1.__awaiter)(void 0, void 0, void 0, function () {
    var customNodes;
    return (0, tslib_1.__generator)(this, function (_a) {
        customNodes = (0, exports.getRegisterNode)(nodes);
        return [2 /*return*/, (0, tslib_1.__spreadArray)((0, tslib_1.__spreadArray)([], customNodes, true), constants_1.NODEPOOL.map(function (_a) {
                var name = _a.name, ports = _a.ports, _b = _a.width, width = _b === void 0 ? constants_1.NODE_WIDTH : _b, _c = _a.height, height = _c === void 0 ? constants_1.NODE_HEIGHT : _c, _d = _a.label, label = _d === void 0 ? '' : _d;
                return {
                    id: (0, xflow_core_1.uuidv4)(),
                    renderKey: name,
                    name: name,
                    label: label,
                    popoverContent: function () { return name; },
                    width: width,
                    height: height,
                    ports: getPorts(ports),
                };
            }), true)];
    });
}); };
exports.nodeService = nodeService;
var setNodeRender = function (config, nodes) {
    if (nodes === void 0) { nodes = []; }
    // 自定义节点
    if (nodes === null || nodes === void 0 ? void 0 : nodes.length) {
        nodes.forEach(function (item) {
            var name = item.name, component = item.component;
            config.setNodeRender(name, component);
        });
    }
    /** 默认节点，通过 Terminal 标识，避免多次调用*/
    if (!config.nodeRender.get('Terminal')) {
        constants_1.NODEPOOL.forEach(function (item) {
            config.setNodeRender(item.name, NodesComponent["".concat(item.name.replace(/\s+/g, ''), "Node")]);
        });
    }
};
exports.setNodeRender = setNodeRender;
// 创建节点路径
var createPath = function (paths, offsetX, offsetY) {
    if (offsetX === void 0) { offsetX = 0; }
    if (offsetY === void 0) { offsetY = 0; }
    if (!paths.length) {
        return null;
    }
    var path = '';
    // @ts-ignore
    paths.forEach(function (item) {
        var c = item[0], x = item[1], y = item[2], c2x = item[3], c2y = item[4];
        path += (0, lodash_1.isNumber)(y) ? " ".concat(c, " ").concat(x + offsetX, " ").concat(y + offsetY) : " ".concat(c);
        if (c2y) {
            path += " ".concat(c2x + offsetX, " ").concat(c2y + offsetY);
        }
    });
    return path;
};
exports.createPath = createPath;
