// @antv/xflow-extension/es/flowchart-canvas/extension/contributions/flow.js
import { __awaiter, __decorate, __rest } from "tslib";
import { singleton } from 'mana-syringe';
import { DisposableCollection, Disposable, XFlowEdgeCommands } from '@antv/xflow-core';
import { IHookContribution } from '@antv/xflow-core';
import { NsAddEdgeEvent } from '../../config-graph';
import { getProps } from '../../utils';
/**
 * 内置的hook contribution
 * 处理 config上的runtime的注册项
 */
let FlowHooksContribution = class FlowHooksContribution {
    constructor() {
        this.toDispose = new DisposableCollection();
        this.registerHookHub = () => __awaiter(this, void 0, void 0, function* () {
            return Disposable.create(() => { });
        });
        this.registerHook = (hooks) => __awaiter(this, void 0, void 0, function* () {
            const toDispose = new DisposableCollection();
            const disposables = [
                hooks.afterGraphInit.registerHook({
                    name: 'call add edge to replace temp edge',
                    handler: (handlerArgs) => __awaiter(this, void 0, void 0, function* () {
                        const { commandService, graph } = handlerArgs;
                        graph.on(NsAddEdgeEvent.EVENT_NAME, (args) => __awaiter(this, void 0, void 0, function* () {
                            const { edge } = args, edgeConfig = __rest(args, ["edge"]);
                            const config = {
                                edgeConfig: Object.assign(Object.assign({}, edgeConfig), { 
                                    // renderKey: FLOWCHART_EDGE, // 暂不支持
                                    source: {
                                        cell: edgeConfig.source,
                                        port: edgeConfig.sourcePortId,
                                    }, target: {
                                        cell: edgeConfig.target,
                                        port: edgeConfig.targetPortId,
                                    }, zIndex: 1, attrs: {
                                        line: {
                                            stroke: '#A2B1C3',
                                            targetMarker: {
                                                name: 'block',
                                                width: 12,
                                                height: 8,
                                            },
                                            strokeDasharray: '5 5',
                                            strokeWidth: 1,
                                        },
                                    }, data: Object.assign({}, edgeConfig) }),
                            };
                            yield commandService.executeCommand(XFlowEdgeCommands.ADD_EDGE.id, config);
                            const onAddEdge = getProps('onAddEdge');
                            if (typeof onAddEdge === 'function') {
                                onAddEdge(config);
                            }
                            args.edge.remove();
                        }));
                    }),
                }),
            ];
            toDispose.pushAll(disposables);
            return Disposable.create(() => { });
        });
    }
};
FlowHooksContribution = __decorate([
    singleton({ contrib: IHookContribution })
], FlowHooksContribution);
export { FlowHooksContribution };
