// @antv/l7-layers/es/imagetile/utils/ImageTile.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import { toLngLatBounds } from '@antv/geo-coord';
import ImageLayer from '../../image';
var r2d = 180 / Math.PI;
var tileURLRegex = /\{([zxy])\}/g;

var ImageTile = function () {
  function ImageTile(key, url, container, sceneContainer) {
    _classCallCheck(this, ImageTile);

    _defineProperty(this, "tile", void 0);

    _defineProperty(this, "name", void 0);

    _defineProperty(this, "imageLayer", void 0);

    this.name = key;
    this.tile = key.split('_').map(function (v) {
      return Number(v);
    });
    var urlParams = {
      x: this.tile[0],
      y: this.tile[1],
      z: this.tile[2]
    };
    var imageSrc = this.getTileURL(urlParams, url);
    var lnglatBounds = this.tileLnglatBounds(this.tile);
    var west = lnglatBounds.getWest();
    var south = lnglatBounds.getSouth();
    var east = lnglatBounds.getEast();
    var north = lnglatBounds.getNorth();
    var imageLayer = new ImageLayer({
      zIndex: -999
    });
    imageLayer.source(imageSrc, {
      parser: {
        type: 'image',
        extent: [west, south, east, north]
      }
    });
    imageLayer.setContainer(container, sceneContainer);
    imageLayer.init();
    this.imageLayer = imageLayer;
  }

  _createClass(ImageTile, [{
    key: "destroy",
    value: function destroy() {
      this.imageLayer.clearModels();
      this.imageLayer.destroy();
    }
  }, {
    key: "getTileURL",
    value: function getTileURL(urlParams, path) {
      if (!urlParams.s) {
        urlParams.s = String.fromCharCode(97 + Math.floor(Math.random() * 3));
      }

      tileURLRegex.lastIndex = 0;
      return path.replace(tileURLRegex, function (value, key) {
        return urlParams[key];
      });
    }
  }, {
    key: "tileLnglatBounds",
    value: function tileLnglatBounds(tile) {
      var e = this.tile2lng(tile[0] + 1, tile[2]);
      var w = this.tile2lng(tile[0], tile[2]);
      var s = this.tile2lat(tile[1] + 1, tile[2]);
      var n = this.tile2lat(tile[1], tile[2]);
      return toLngLatBounds([w, n], [e, s]);
    }
  }, {
    key: "tile2lng",
    value: function tile2lng(x, z) {
      return x / Math.pow(2, z) * 360 - 180;
    }
  }, {
    key: "tile2lat",
    value: function tile2lat(y, z) {
      var n = Math.PI - 2 * Math.PI * y / Math.pow(2, z);
      return r2d * Math.atan(0.5 * (Math.exp(n) - Math.exp(-n)));
    }
  }]);

  return ImageTile;
}();

export { ImageTile as default };
