// @antv/l7-layers/es/imagetile/utils/tileCache.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import LRUCache from './lruCache';

var TileCache = function () {
  function TileCache() {
    var limit = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 50;
    var tileDestroy = arguments.length > 1 ? arguments[1] : undefined;

    _classCallCheck(this, TileCache);

    _defineProperty(this, "cache", void 0);

    this.cache = new LRUCache(limit, tileDestroy);
  }

  _createClass(TileCache, [{
    key: "getTile",
    value: function getTile(key) {
      return this.cache.get(key);
    }
  }, {
    key: "setTile",
    value: function setTile(tile, key) {
      this.cache.set(key, tile);
    }
  }, {
    key: "destory",
    value: function destory() {
      this.cache.clear();
    }
  }, {
    key: "setNeedUpdate",
    value: function setNeedUpdate() {
      var _this = this;

      this.cache.order.forEach(function (key) {
        var tile = _this.cache.get(key);

        tile.needUpdate = true;
      });
    }
  }]);

  return TileCache;
}();

export { TileCache as default };
