// @ant-design/flowchart/lib/components/node-panel/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ASPECTRATIONODE = exports.NODEPOOL = exports.NODE_PADDING = exports.NODE_HEIGHT = exports.NODE_WIDTH = void 0;
// 节点默认尺寸
exports.NODE_WIDTH = 60;
exports.NODE_HEIGHT = 40;
// svg 绘制起始点，解决边不清晰的问题
exports.NODE_PADDING = 1;
exports.NODEPOOL = [
    {
        name: 'Terminal',
    },
    {
        name: 'Process',
    },
    {
        name: 'Decision',
    },
    {
        name: 'Multi Document',
    },
    {
        name: 'Connector',
        width: exports.NODE_HEIGHT,
        height: exports.NODE_HEIGHT,
    },
    {
        name: 'DataIO',
        ports: ['top', 'bottom'],
    },
    {
        name: 'Database',
    },
    {
        name: 'Hard Disk',
    },
    {
        name: 'Stroed Data',
    },
    {
        name: 'Document',
    },
    {
        name: 'Predefined Process',
    },
    {
        name: 'Extract',
    },
    {
        name: 'Merge',
    },
    {
        name: 'Or',
        width: exports.NODE_HEIGHT,
        height: exports.NODE_HEIGHT,
    },
    {
        name: 'Manual Input',
    },
    {
        name: 'Preparation',
    },
    {
        name: 'Delay',
    },
    {
        name: 'Manual Operation',
    },
    {
        name: 'Display',
    },
    {
        name: 'Off Page Link',
    },
    {
        name: 'Note Left',
        label: '≣',
    },
    {
        name: 'Note Right',
        label: '≣',
    },
    {
        name: 'Internal Storage',
    },
    {
        name: 'Text',
        label: 'text',
    },
];
// 缩放时保存同比例
exports.ASPECTRATIONODE = ['Connector', 'Or'];
