// @antv/xflow-extension/lib/canvas-scale-toolbar/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useConfig = exports.TOOLBAR_TYPE = exports.CANVAS_SCALE_TOOLBAR_CONFIG = void 0;
var tslib_1 = require("tslib");
var xflow_core_1 = require("@antv/xflow-core");
var icons_1 = require("@ant-design/icons");
var canvas_toolbar_1 = require("../canvas-toolbar");
var CANVAS_SCALE_TOOLBAR_CONFIG;
(function (CANVAS_SCALE_TOOLBAR_CONFIG) {
    xflow_core_1.IconStore.set('ZoomInOutlined', icons_1.ZoomInOutlined);
    xflow_core_1.IconStore.set('ZoomOutOutlined', icons_1.ZoomOutOutlined);
    xflow_core_1.IconStore.set('OneToOneOutlined', icons_1.OneToOneOutlined);
    xflow_core_1.IconStore.set('CompressOutlined', icons_1.CompressOutlined);
    CANVAS_SCALE_TOOLBAR_CONFIG.ZOOM_IN = xflow_core_1.XFlowGraphCommands.GRAPH_ZOOM.id + '-zoom-in';
    CANVAS_SCALE_TOOLBAR_CONFIG.ZOOM_OUT = xflow_core_1.XFlowGraphCommands.GRAPH_ZOOM.id + '-zoom-out';
    CANVAS_SCALE_TOOLBAR_CONFIG.SCALE_TO_ONE = xflow_core_1.XFlowGraphCommands.GRAPH_ZOOM.id + '-scale-to-one';
    CANVAS_SCALE_TOOLBAR_CONFIG.SCALE_TO_FIT = xflow_core_1.XFlowGraphCommands.GRAPH_ZOOM.id + '-scale-to-fit';
    CANVAS_SCALE_TOOLBAR_CONFIG.MAX_SCALE = 1.5;
    CANVAS_SCALE_TOOLBAR_CONFIG.MIN_SCALE = 0.05;
    CANVAS_SCALE_TOOLBAR_CONFIG.zoomOptions = {
        maxScale: CANVAS_SCALE_TOOLBAR_CONFIG.MAX_SCALE,
        minScale: CANVAS_SCALE_TOOLBAR_CONFIG.MIN_SCALE,
    };
    CANVAS_SCALE_TOOLBAR_CONFIG.getToolbarConfig = function (zoomFactor) {
        return [
            {
                name: 'main',
                items: [
                    {
                        id: CANVAS_SCALE_TOOLBAR_CONFIG.ZOOM_IN,
                        tooltip: '放大',
                        iconName: 'ZoomInOutlined',
                        onClick: function (_a) {
                            var commandService = _a.commandService;
                            commandService.executeCommand(xflow_core_1.XFlowGraphCommands.GRAPH_ZOOM.id, {
                                factor: 0.1,
                                zoomOptions: CANVAS_SCALE_TOOLBAR_CONFIG.zoomOptions,
                            });
                        },
                    },
                    {
                        id: CANVAS_SCALE_TOOLBAR_CONFIG.ZOOM_OUT,
                        tooltip: '缩小',
                        iconName: 'ZoomOutOutlined',
                        onClick: function (_a) {
                            var commandService = _a.commandService;
                            commandService.executeCommand(xflow_core_1.XFlowGraphCommands.GRAPH_ZOOM.id, {
                                factor: -0.1,
                                zoomOptions: CANVAS_SCALE_TOOLBAR_CONFIG.zoomOptions,
                            });
                        },
                    },
                    {
                        id: CANVAS_SCALE_TOOLBAR_CONFIG.SCALE_TO_ONE,
                        iconName: 'OneToOneOutlined',
                        tooltip: '缩放到1:1',
                        isEnabled: zoomFactor !== 1,
                        onClick: function (_a) {
                            var commandService = _a.commandService;
                            commandService.executeCommand(xflow_core_1.XFlowGraphCommands.GRAPH_ZOOM.id, {
                                factor: 'real',
                                zoomOptions: CANVAS_SCALE_TOOLBAR_CONFIG.zoomOptions,
                            });
                        },
                    },
                    {
                        id: CANVAS_SCALE_TOOLBAR_CONFIG.SCALE_TO_FIT,
                        tooltip: '缩放到适应屏幕',
                        iconName: 'CompressOutlined',
                        onClick: function (_a) {
                            var commandService = _a.commandService;
                            commandService.executeCommand(xflow_core_1.XFlowGraphCommands.GRAPH_ZOOM.id, {
                                factor: 'fit',
                                zoomOptions: CANVAS_SCALE_TOOLBAR_CONFIG.zoomOptions,
                            });
                        },
                    },
                ],
            },
        ];
    };
})(CANVAS_SCALE_TOOLBAR_CONFIG = exports.CANVAS_SCALE_TOOLBAR_CONFIG || (exports.CANVAS_SCALE_TOOLBAR_CONFIG = {}));
exports.TOOLBAR_TYPE = 'CANVAS_SCALE_TOOLBAR';
exports.useConfig = (0, canvas_toolbar_1.createToolbarConfig)(function (config) {
    config.setToolbarModelService(function (model, modelService) { return (0, tslib_1.__awaiter)(void 0, void 0, void 0, function () {
        var graphScale, graphScaleModel;
        return (0, tslib_1.__generator)(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, xflow_core_1.MODELS.GRAPH_SCALE.useValue(modelService)
                    /** 设置初始值*/
                ];
                case 1:
                    graphScale = _a.sent();
                    /** 设置初始值*/
                    model.setValue(function (m) {
                        m.mainGroups = CANVAS_SCALE_TOOLBAR_CONFIG.getToolbarConfig(graphScale.zoomFactor);
                    });
                    return [4 /*yield*/, xflow_core_1.MODELS.GRAPH_SCALE.getModel(modelService)
                        /** graphScaleModel更新时联动 Toolbar*/
                    ];
                case 2:
                    graphScaleModel = _a.sent();
                    /** graphScaleModel更新时联动 Toolbar*/
                    graphScaleModel.watch(function (_a) {
                        var zoomFactor = _a.zoomFactor;
                        model.setValue(function (m) {
                            m.mainGroups = CANVAS_SCALE_TOOLBAR_CONFIG.getToolbarConfig(zoomFactor);
                        });
                    });
                    return [2 /*return*/];
            }
        });
    }); });
});
