// @antv/l7-utils/lib/mini-adapter/blob.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Blob = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var Blob = function () {
  function Blob(buffers, type) {
    (0, _classCallCheck2.default)(this, Blob);
    this.buffers = buffers;
    this.type = type;
  }

  (0, _createClass2.default)(Blob, [{
    key: "arraybuffer",
    value: function arraybuffer() {
      return Promise.resolve(this.buffers[0]);
    }
  }, {
    key: "stream",
    value: function stream() {
      throw new Error('not implemented');
    }
  }, {
    key: "text",
    value: function text() {
      throw new Error('not implemented');
    }
  }, {
    key: "slice",
    value: function slice(start, end, contentType) {
      throw new Error('not implemented');
    }
  }]);
  return Blob;
}();

exports.Blob = Blob;
