// @antv/l7-utils/lib/mini-adapter/document.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.$document = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _Event = require("./Event");

var _HTMLElement3 = require("./HTMLElement");

var _HTMLVideoElement = require("./HTMLVideoElement");

var _Image = require("./Image");

var _location = require("./location");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var Body = function (_HTMLElement) {
  (0, _inherits2.default)(Body, _HTMLElement);

  var _super = _createSuper(Body);

  function Body() {
    (0, _classCallCheck2.default)(this, Body);
    return _super.call(this, 'body', 0);
  }

  (0, _createClass2.default)(Body, [{
    key: "addEventListener",
    value: function addEventListener(type, listener) {
      var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
    }
  }, {
    key: "removeEventListener",
    value: function removeEventListener(type, listener, options) {}
  }, {
    key: "dispatchEvent",
    value: function dispatchEvent(event) {}
  }]);
  return Body;
}(_HTMLElement3.HTMLElement);

var DocumentElement = function (_HTMLElement2) {
  (0, _inherits2.default)(DocumentElement, _HTMLElement2);

  var _super2 = _createSuper(DocumentElement);

  function DocumentElement() {
    (0, _classCallCheck2.default)(this, DocumentElement);
    return _super2.call(this, 'html', 0);
  }

  (0, _createClass2.default)(DocumentElement, [{
    key: "addEventListener",
    value: function addEventListener(type, listener) {
      var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
    }
  }, {
    key: "removeEventListener",
    value: function removeEventListener(type, listener) {}
  }, {
    key: "dispatchEvent",
    value: function dispatchEvent(event) {}
  }]);
  return DocumentElement;
}(_HTMLElement3.HTMLElement);

var events = {};
var $document = {
  readyState: 'complete',
  visibilityState: 'visible',
  hidden: false,
  fullscreen: true,
  scripts: [],
  style: {},
  location: _location.$location,
  ontouchstart: null,
  ontouchmove: null,
  ontouchend: null,
  onvisibilitychange: null,
  parentNode: null,
  parentElement: null,
  head: null,
  body: null,
  documentElement: {
    style: []
  },
  createElement: function createElement(tagName) {
    tagName = tagName.toLowerCase();

    if (tagName === 'canvas') {
      return my.createOffscreenCanvas(1024, 128, '2d');
    } else if (tagName === 'img') {
      return new _Image.Image();
    } else if (tagName === 'video') {
      return new _HTMLVideoElement.HTMLVideoElement();
    }

    return new _HTMLElement3.HTMLElement(tagName);
  },
  createElementNS: function createElementNS(nameSpace, tagName) {
    return this.createElement(tagName);
  },
  createTextNode: function createTextNode(text) {
    return text;
  },
  getElementById: function getElementById(id) {
    return null;
  },
  getElementsByTagName: function getElementsByTagName(tagName) {
    tagName = tagName.toLowerCase();
    return [];
  },
  getElementsByTagNameNS: function getElementsByTagNameNS(nameSpace, tagName) {
    return this.getElementsByTagName(tagName);
  },
  getElementsByName: function getElementsByName(tagName) {
    return [];
  },
  querySelector: function querySelector(query) {
    return null;
  },
  querySelectorAll: function querySelectorAll(query) {
    return [];
  },
  addEventListener: function addEventListener(type, listener, options) {
    if (!events[type]) {
      events[type] = [];
    }

    events[type].push(listener);
  },
  removeEventListener: function removeEventListener(type, listener) {
    var listeners = events[type];

    if (listeners && listeners.length > 0) {
      for (var i = listeners.length; i--; i > 0) {
        if (listeners[i] === listener) {
          listeners.splice(i, 1);
          break;
        }
      }
    }
  },
  dispatchEvent: function dispatchEvent(event) {
    var type = event.type;
    var listeners = events[type];

    if (listeners) {
      for (var i = 0; i < listeners.length; i++) {
        listeners[i](event);
      }
    }

    if (event.target && typeof event.target['on' + type] === 'function') {
      event.target['on' + type](event);
    }
  },
  createEvent: function createEvent(type) {
    return new _Event.Event(type);
  }
};
exports.$document = $document;
$document.documentElement = new DocumentElement();
$document.head = new _HTMLElement3.HTMLElement('head');
$document.body = new Body();
