// @antv/l7-utils/lib/mini-adapter/Event.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Event = void 0;

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var Event = (0, _createClass2.default)(function Event(type, data) {
  (0, _classCallCheck2.default)(this, Event);
  (0, _defineProperty2.default)(this, "cancelBubble", void 0);
  (0, _defineProperty2.default)(this, "cancelable", void 0);
  (0, _defineProperty2.default)(this, "target", void 0);
  (0, _defineProperty2.default)(this, "currentTarget", void 0);
  (0, _defineProperty2.default)(this, "preventDefault", void 0);
  (0, _defineProperty2.default)(this, "stopPropagation", void 0);
  (0, _defineProperty2.default)(this, "type", void 0);
  (0, _defineProperty2.default)(this, "timeStamp", void 0);
  this.cancelBubble = false;
  this.cancelable = false;
  this.target = null;
  this.currentTarget = null;

  this.preventDefault = function () {};

  this.stopPropagation = function () {};

  this.type = type;
  this.timeStamp = Date.now();
});
exports.Event = Event;
