// @antv/l7-utils/lib/mini-adapter/screen.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.screen = exports.isMiniAli = void 0;

var _global = require("./global");

var isMiniAli = typeof my !== 'undefined' && !!my && typeof my.showToast === 'function' && my.isFRM !== true;
exports.isMiniAli = isMiniAli;
var screenWidth;
var screenHeight;
var windowWidth;
var windowHeight;

if (isMiniAli) {
  var myOptions = my.getSystemInfoSync();
  screenWidth = myOptions.screenWidth;
  screenHeight = myOptions.screenHeight;
  windowWidth = myOptions.windowWidth;
  windowHeight = myOptions.windowHeight;
} else {
  var innerWidth = _global.globalWindow.innerWidth,
      innerHeight = _global.globalWindow.innerHeight;
  screenWidth = innerWidth;
  screenHeight = innerHeight;
  windowWidth = innerWidth;
  windowHeight = innerHeight;
}

var screen = {
  width: screenWidth,
  height: screenHeight,
  availWidth: windowWidth,
  availHeight: windowHeight,
  availLeft: 0,
  availTop: 0
};
exports.screen = screen;
