// @antv/l7-maps/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Earth", {
  enumerable: true,
  get: function get() {
    return _earth.default;
  }
});
Object.defineProperty(exports, "GaodeMap", {
  enumerable: true,
  get: function get() {
    return _amap.default;
  }
});
Object.defineProperty(exports, "GaodeMapV2", {
  enumerable: true,
  get: function get() {
    return _amap2.default;
  }
});
Object.defineProperty(exports, "Map", {
  enumerable: true,
  get: function get() {
    return _map.default;
  }
});
Object.defineProperty(exports, "Mapbox", {
  enumerable: true,
  get: function get() {
    return _mapbox.default;
  }
});
Object.defineProperty(exports, "Version", {
  enumerable: true,
  get: function get() {
    return _version.Version;
  }
});

var _amap = _interopRequireDefault(require("./amap/"));

var _amap2 = _interopRequireDefault(require("./amap2/"));

var _earth = _interopRequireDefault(require("./earth/"));

var _map = _interopRequireDefault(require("./map/"));

var _mapbox = _interopRequireDefault(require("./mapbox/"));

var _version = require("./version");
