// @antv/geo-coord/lib/geo/projection/mercator.js
"use strict";
exports.__esModule = true;
var lng_lat_1 = require("../geometry/lng-lat");
var bounds_1 = require("../geometry/bounds");
var point_1 = require("../geometry/point");
exports.mercator = {
    R: 6378137,
    R_MINOR: 6356752.314245179,
    bounds: new bounds_1.Bounds([-20037508.34279, -15496570.73972], [20037508.34279, 18764656.23138]),
    project: function (lnglat) {
        var d = Math.PI / 180;
        var r = this.R;
        var tmp = this.R_MINOR / r;
        var e = Math.sqrt(1 - tmp * tmp);
        var y = lnglat.lat * d;
        var con = e * Math.sin(y);
        var ts = Math.tan(Math.PI / 4 - y / 2) / Math.pow((1 - con) / (1 + con), e / 2);
        y = -r * Math.log(Math.max(ts, 1e-10));
        return new point_1.Point(lnglat.lng * d * r, y);
    },
    unproject: function (point) {
        var d = 180 / Math.PI;
        var r = this.R;
        var tmp = this.R_MINOR / r;
        var e = Math.sqrt(1 - tmp * tmp);
        var ts = Math.exp(-point.y / r);
        var phi = Math.PI / 2 - 2 * Math.atan(ts);
        for (var i = 0, dphi = 0.1, con = void 0; i < 15 && Math.abs(dphi) > 1e-7; i += 1) {
            con = e * Math.sin(phi);
            con = Math.pow((1 - con) / (1 + con), e / 2);
            dphi = Math.PI / 2 - 2 * Math.atan(ts * con) - phi;
            phi += dphi;
        }
        return new lng_lat_1.LngLat((point.x * d) / r, phi * d);
    }
};
